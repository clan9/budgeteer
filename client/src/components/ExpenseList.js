import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getExpenses } from '../actions/expenses';
import selectExpenses from '../selectors/selectExpenses';
import ExpenseListItem from './ExpenseListItem';

export class ExpenseList extends Component {
  componentDidMount() {
    this.props.getExpenses();
  }

  render() {
    return (
      <div className="content-container">
        <div className="list-header">
          <div className="hide-on-desktop">Expenses</div>
          <div className="hide-on-mobile">Expenses</div>
          <div className="hide-on-mobile">Amount</div>
        </div>
        <div className="list-body">
          {this.props.expenses.length === 0 ? (
            <div className="list-item list-item--message">
              <span>No expenses</span>
            </div>
          ) : (
            this.props.expenses.map(expense => {
              return <ExpenseListItem key={expense._id} {...expense} />;
            })
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(
  mapStateToProps,
  { getExpenses }
)(ExpenseList);
/*

<div className="content-container">
  <div className="list-header">
    <div className="show-for-mobile">Expenses</div>
    <div className="show-for-desktop">Expense</div>
    <div className="show-for-desktop">Amount</div>
  </div>
  <div className="list-body">
    {props.expenses.length === 0 ? (
      <div className="list-item list-item--message">
        <span>No expenses</span>
      </div>
    ) : (
      props.expenses.map(expense => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
</div>;

*/
