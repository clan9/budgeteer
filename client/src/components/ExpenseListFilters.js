import 'react-dates/initialize';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  setStartDate,
  setEndDate,
  setTextFilter,
  sortByAmount,
  sortByDate
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends Component {
  state = { calendarFocused: null };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    e.target.value === 'date'
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search Expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDateId="start"
              endDateId="end"
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(
  mapStateToProps,
  {
    setStartDate,
    setEndDate,
    setTextFilter,
    sortByAmount,
    sortByDate
  }
)(ExpenseListFilters);
