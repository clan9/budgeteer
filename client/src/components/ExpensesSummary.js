import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/selectExpenses';
import selectedExpensesTotal from '../selectors/selectedExpensesTotal';

require('numeral/locales/en-gb');
numeral.locale('en-gb');

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expenseTotal / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCount}</span> {expenseWord} totalling{' '}
          <span>{formattedExpensesTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: selectedExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
