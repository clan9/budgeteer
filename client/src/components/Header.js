import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => {
  return (
    <header className="header">
      <div className="content-container">
        <div className="header__content">
          <Link className="header__title" to="/dash">
            <h1>Budgeteer</h1>
          </Link>
          <a href="/auth/logout" className="button button--link">
            Logout
          </a>
        </div>
      </div>
    </header>
  );
};

export default Header;
