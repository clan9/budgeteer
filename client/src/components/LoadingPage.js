import React from 'react';

const LoadingPage = () => (
  <div className="loader">
    <img
      className="loader__image"
      src="/images/loader.gif"
      alt="Page Loading"
    />
  </div>
);

export default LoadingPage;
