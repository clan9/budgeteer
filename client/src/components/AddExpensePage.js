import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import Header from './Header';

export const AddExpensePage = ({ addExpense, history, user, auth }) => {
  if (!user && !auth) {
    return <Redirect to="/" />;
  } else {
    return (
      <div>
        <Header />
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add an Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit={expense => {
              addExpense(expense);
              history.push('/dash');
            }}
          />
        </div>
      </div>
    );
  }
};

const mapStateToProps = state => ({
  user: state.auth.user,
  loading: state.auth.loading,
  auth: state.auth.isAuthenticated
});

export default connect(
  mapStateToProps,
  { addExpense }
)(AddExpensePage);
