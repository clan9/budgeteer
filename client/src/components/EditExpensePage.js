import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import Header from './Header';

export const EditExpensePage = props => {
  if (!props.user && !props.auth) {
    return <Redirect to="/" />;
  } else {
    return (
      <div>
        <Header />
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={props.expense}
            onSubmit={expense => {
              props.editExpense(props.expense._id, expense);
              props.history.push('/dash');
            }}
          />
          <button
            className="button button--secondary"
            onClick={e => {
              props.removeExpense(props.expense._id);
              props.history.push('/dash');
            }}
          >
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find(
    expense => expense._id === ownProps.match.params.id
  ),
  user: state.auth.user,
  auth: state.auth.isAuthenticated
});

export default connect(
  mapStateToProps,
  { editExpense, removeExpense }
)(EditExpensePage);
