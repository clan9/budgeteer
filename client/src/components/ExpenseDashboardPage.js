import React, { Fragment, Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { fetchUser } from '../actions/auth';
import Header from './Header';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseList from './ExpenseList';
import LoadingPage from './LoadingPage';

export class ExpenseDashboardPage extends Component {
  componentDidMount() {
    this.props.fetchUser();
  }

  renderContent() {
    if (this.props.loading) {
      return <LoadingPage />;
    } else if (!this.props.loading && !this.props.user) {
      return <Redirect to="/" />;
    } else {
      return (
        <Fragment>
          <Header />
          <ExpensesSummary />
          <ExpenseListFilters />
          <ExpenseList />
        </Fragment>
      );
    }
  }

  render() {
    return <div>{this.renderContent()}</div>;
  }
}

const mapStateToProps = state => ({
  user: state.auth.user,
  isAuthenticated: state.auth.isAuthenticated,
  loading: state.auth.loading
});

export default connect(
  mapStateToProps,
  { fetchUser }
)(ExpenseDashboardPage);
