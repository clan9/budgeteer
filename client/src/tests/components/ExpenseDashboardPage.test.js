import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseDashboardPage } from '../../components/ExpenseDashboardPage';

describe('Expense Dashboard test', () => {
  let fetchUser;

  beforeEach(() => {
    fetchUser = jest.fn();
  });

  it('should correctly render Expense dashboard when logged in', () => {
    const loading = false;
    const user = true;
    const wrapper = shallow(
      <ExpenseDashboardPage
        fetchUser={fetchUser}
        loading={loading}
        user={user}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect if no valid user found', () => {
    const loading = false;
    const user = false;
    const wrapper = shallow(
      <ExpenseDashboardPage
        fetchUser={fetchUser}
        loading={loading}
        user={user}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should show LoadingPage if Dashboard not yet loaded', () => {
    const loading = true;
    const wrapper = shallow(
      <ExpenseDashboardPage fetchUser={fetchUser} loading={loading} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
