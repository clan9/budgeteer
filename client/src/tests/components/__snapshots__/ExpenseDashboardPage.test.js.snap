// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense Dashboard test should correctly render Expense dashboard when logged in 1`] = `
<div>
  <Header />
  <Connect(ExpensesSummary) />
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;

exports[`Expense Dashboard test should redirect if no valid user found 1`] = `
<div>
  <Redirect
    to="/"
  />
</div>
`;

exports[`Expense Dashboard test should show LoadingPage if Dashboard not yet loaded 1`] = `
<div>
  <LoadingPage />
</div>
`;
